test/test-parse-warning.cpp

```diff
@@ -672,6 +672,68 @@ TEST_F(test_parse_warning,
         }));
   }
 }
+
+
+TEST_F(test_parse_warning,
+       warn_on_pointless_nullish_coalescing_operator) {
+  
+  {
+    test_parser p(u8"true ?? false"_sv, capture_diags);
+    p.parse_and_visit_expression();
+
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(
+                p.code, diag_pointless_nullish_coalescing_operator,
+                question_question, strlen(u8"true "), u8"??"_sv),
+        }));
+  }
+  {
+    test_parser p(u8"x == y ?? true"_sv, capture_diags);
+    p.parse_and_visit_expression();
+
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(
+                p.code, diag_pointless_nullish_coalescing_operator,
+                question_question, strlen(u8"x == y "), u8"??"_sv),
+        }));
+  }
+  {
+    test_parser p(u8"!b ?? y != x"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(
+                p.code, diag_pointless_nullish_coalescing_operator,
+                question_question, strlen(u8"!b "), u8"??"_sv),
+        }));
+  }
+  {
+    test_parser p(u8"'hi' ?? true"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    EXPECT_THAT(
+        p.errors,
+        ElementsAreArray({
+            DIAG_TYPE_OFFSETS(
+                p.code, diag_pointless_nullish_coalescing_operator,
+                question_question, strlen(u8"'hi' "), u8"??"_sv),
+        }));
+  }
+  {
+    test_parser p(u8"s.toLowerCase() ?? false"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    EXPECT_THAT(p.errors, IsEmpty());
+  }
+  {
+    test_parser p(u8"s ?? false"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    EXPECT_THAT(p.errors, IsEmpty());
+  }
```

Comment:
Nitpick: You can make these tests shorter if you remove `capture_diags`:
```suggestion
  {
    test_parser p(u8"s.toLowerCase() ?? false"_sv);
    p.parse_and_visit_expression();
  }
  {
    test_parser p(u8"s ?? false"_sv);
    p.parse_and_visit_expression();
  }
```

Reply:
I removed those and the tests passed the same so you're definitely right. So is the capture_diags parameter only necessary for false negative testing? I ask since other tests in this file that test for IsEmpty() include capture_diags. I also see that you deleted EXPECT_THAT(p.errors, IsEmpty()); I thought that was neccessary too, but it seems to work exactly the same without it, so I can delete those statements as well. I only ask since I based my test cases off of other test cases in this file which follow this format.

Reply:
> So is the capture_diags parameter only necessary for false negative testing?

If you add `capture_diags`, then `p.errors` is filled with whatever diagnostics are emitted by the code. You can then write an assertion for `p.errors`.

If you do not add `capture_diags`, then any diagnostic will cause the test to fail.

> I ask since other tests in this file that test for IsEmpty() include capture_diags. [...] I only ask since I based my test cases off of other test cases in this file which follow this format.

Yeah, those tests must be old and not written in the new style.